[Beta] docs(cn): translate scaling-up-with-reducer-and-context (#694)
reactjs/zh-hans.reactjs.org: beta/src/pages/learn/scaling-up-with-reducer-and-context.md

---

> title: 使用 Reducer 和 Context 来拓展你的应用

Reviewer: ```suggestions
title: 使用 Reducer 和 Context 来拓展你的应用
translator:
 - Ikaite
 - KnowsCount
 - TinaaaaP
 - yyyang1996
 ```

> title: 使用 Reducer 和 Context 来拓展你的应用

Reviewer: ```suggestions
title: 使用 Reducer 和 Context 来拓展你的应用
translator: 
  - Ikaite 
  - KnowsCount 
  - TinaaaaP 
  - yyyang1996
```

---
title: 使用 Reducer 和 Context 来拓展你的应用

> translator: 

Reviewer 2: ```suggestion
translators: 
```

---
title: 使用 Reducer 和 Context 来拓展你的应用
---

<Intro>

Reducer 可以整合组件的状态更新逻辑。Context 可以将信息深入传递给其他组件。你可以组合使用它们来共同管理一个复杂页面的状态。

</Intro>

<YouWillLearn>

* 如何结合使用 reducer 和 context
* 如何去避免通过 props 传递 state 和 dispatch
* 如何将 context 和状态逻辑保存在一个单独的文件中

</YouWillLearn>

## 结合使用 reducer 和 context {/*combining-a-reducer-with-context*/}

> 在 [reducer 介绍](/learn/extracting-state-logic-into-a-reducer)的例子里面，状态被reducer所管理。reducer函数包含了所有的状态更新逻辑并在此文件的底部声明：

Reviewer: ```suggestion
在 [reducer 介绍](/learn/extracting-state-logic-into-a-reducer) 的例子里面，状态被 reducer 所管理。reducer 函数包含了所有的状态更新逻辑并在此文件的底部声明：
```

</Sandpack>

> Reducer 有助于保持事件处理程序的简短明了。但随着应用规模越来越庞大，你就可能会遇到别的困难。**目前，`tasks` 状态和 `dispatch` 函数仅在顶级 `TaskApp` 组件中可用。** 要让其他组件读取任务列表或更改它，你必须显式 [传递](/learn/passing-props-to-a-component) 当前状态和将其更改为 props 的事件处理程序。

Reviewer 3: ```suggestion
Reducer 有助于保持事件处理程序的简短明了。但随着应用规模越来越庞大，你就可能会遇到别的困难。**目前，`tasks` 状态和 `dispatch` 函数仅在顶级 `TaskApp` 组件中可用**。要让其他组件读取任务列表或更改它，你必须显式 [传递](/learn/passing-props-to-a-component) 当前状态和将其更改为 props 的事件处理程序。
```

</Sandpack>

> reducer 有助于事件处理程序简短明了。但随着应用越来越大，你有可能会遇到别的困难。 **目前，`tasks` 状态和 `dispatch` 函数仅在顶级 `TaskBoard` 组件中可用** 若要让其他组件读取 `tasks`状态或对其进行更改，必须以 props 的形式显式 [传递](/learn/passing-props-to-a-component) 它。

Reviewer 3: ```suggestion
Reducer 有助于事件处理程序简短明了。但随着应用越来越规模庞大，你就可能会遇到别的困难。 **目前，`tasks` 状态和 `dispatch` 函数仅在顶级 `TaskBoard` 组件中可用。** 若要让其他组件读取 `tasks`状态或对其进行更改，必须以 props 的形式显式 [传递](/learn/passing-props-to-a-component) 它。
```

> Reducer 有助于事件处理程序简短明了。但随着应用规模越来越庞大，你就可能会遇到别的困难。 **目前，`tasks` 状态和 `dispatch` 函数仅在顶级 `TaskBoard` 组件中可用。** 若要让其他组件读取 `tasks`状态或对其进行更改，必须以 props 的形式显式 [传递](/learn/passing-props-to-a-component) 它。

Reviewer 3: ```suggestion
Reducer 有助于保持事件处理程序的简短明了。但随着应用规模越来越庞大，你就可能会遇到别的困难。 **目前，`tasks` 状态和 `dispatch` 函数仅在顶级 `TaskBoard` 组件中可用。** 若要让其他组件读取 `tasks` 状态或对其进行更改，必须以 props 的形式显式 [传递](/learn/passing-props-to-a-component) 它。
```

> Reducer 有助于保持事件处理程序的简短明了。但随着应用规模越来越庞大，你就可能会遇到别的困难。 **目前，`tasks` 状态和 `dispatch` 函数仅在顶级 `TaskBoard` 组件中可用。** 若要让其他组件读取 `tasks` 状态或对其进行更改，必须以 props 的形式显式 [传递](/learn/passing-props-to-a-component) 它。

Reviewer 3: ```suggestion
Reducer 有助于保持事件处理程序的简短明了。但随着应用规模越来越庞大，你就可能会遇到别的困难。**目前，`tasks` 状态和 `dispatch` 函数仅在顶级 `TaskBoard` 组件中可用。** 若要让其他组件读取 `tasks` 状态或对其进行更改，必须以 props 的形式显式 [传递](/learn/passing-props-to-a-component) 它。
```

/>
```

> 在这样的小例子里他可能问题不大，但是如果你有成千上百个组件时，传递所有状态和函数就会让你非常烦恼！

Reviewer: ```suggestion
在像这样的小示例里这样做没什么问题，但是如果你有成千上百个组件，传递所有状态和函数就会让你非常烦恼！
```

> 在像这样的小示例里这样做没什么问题，但是如果你有成千上百个组件，传递所有状态和函数就会让你非常烦恼！

Reviewer 4: ```suggestion
在像这样的小示例里这样做没什么问题，但是如果你有成千上百个组件，传递所有状态和函数可能会非常麻烦！
```

> 他能工作的非常好在诸如这样的小例子中，但是如果你有成千上百个组件时，传递所有状态和函数会让人非常烦恼！

Reviewer 3: ```suggestion
在这样的小例子里他可能问题不大，但是如果你有成千上百个组件时，传递所有状态和函数就会让你非常烦恼！
```

<!--(TODO: illustration of prop drilling)-->

> 这就是为什么，比起通过 props 传递它们，你可能想把 `tasks` 状态和 `dispatch` 函数都 [放入 context](/learn/passing-data-deeply-with-context)。**这样，所有的在 `TaskBoard` 组件树之下的组件都不必一直往下传 props 而可以直接读取 tasks 和 dispatch 函数。**

Reviewer 3: ```suggestion
这就是为什么，比起通过 props 传递它们，你可能想把 `tasks` 状态和 `dispatch` 函数都 [放入 context](/learn/passing-data-deeply-with-context)。**这样，所有的在 `TaskBoard` 组件树之下的组件都不必一直往下传 props 而可以直接读取 tasks 和 dispatch 函数**。
```

<!--(TODO: illustration of context)-->

下面将介绍如何结合使用 reducer 和 context：

1. **创建** context。
2. 将 state 和 dispatch **放入** context。
3. 在组件树的任何地方 **使用** context。

### 第一步: 创建 context {/*step-1-create-the-context*/}

`useReducer` 返回当前的 `tasks` 和 `dispatch` 函数来让你更新它们：

```js
const [tasks, dispatch] = useReducer(tasksReducer, initialTasks);
```

> 为了将他们从组件树往下传，你将 [创建](/learn/passing-data-deeply-with-context#step-2-use-the-context) 两个不同的 context：

Reviewer 3: ```suggestion
为了将它们从组件树往下传，你将 [创建](/learn/passing-data-deeply-with-context#step-2-use-the-context) 两个不同的 context：
```

/>
```

> `TaskList` 将事件处理器传递给 `Task`：

Reviewer: ```suggestion
`TaskList` 将事件处理程序传递给 `Task`：
```
和上文保持统一，event handlers 都译为事件处理程序

</Sandpack>

> 在这里，您将 `null` 作为默认值传递给两个 context。实际值将由 `TaskBoard` 组件提供。

Reviewer 3: ```suggestion
在这里，你把 `null` 作为默认值传递给两个 context。实际值是由 `TaskBoard` 组件提供的。
```

### 第二步: 将 state 和 dispatch 函数 放入 context {/*step-2-put-state-and-dispatch-into-context*/}

> 现在，您可以将所有的 context 导入 `TaskBoard` 组件。获取 `useReducer()` 返回的 `tasks` 和 `dispatch` 并将它们 [提供](/learn/passing-data-deeply-with-context#step-3-provide-the-context) 给整个组件树：

Reviewer 3: ```suggestion
现在，你可以将所有的 context 导入 `TaskBoard` 组件。获取 `useReducer()` 返回的 `tasks` 和 `dispatch` 并将它们 [提供](/learn/passing-data-deeply-with-context#step-3-provide-the-context) 给整个组件树：
```

```

```css
button {
    margin: 5px;
}

li {
    list-style-type: none;
}

ul, li {
    margin: 0;
    padding: 0;
}
```

</Sandpack>

在这里，您将 `null` 作为默认值传递给两个 context。实际值将由 `TaskBoard` 组件提供。

### 第二步: 将 state 和 dispatch 函数 放入context {/*step-2-put-state-and-dispatch-into-context*/}

现在，您可以将所有的 context 导入 `TaskBoard` 组件。获取 `useReducer()` 返回的 `tasks` 和 `dispatch`

> 并将它们[提供](/learn/passing-data-deeply-with-context#step-3-provide-the-context)给整个组件树：

Reviewer: ```suggestion
并将它们 [提供](/learn/passing-data-deeply-with-context#step-3-provide-the-context) 给整个组件树：
```

}
```

> 现在，您可以同时通过 props 和 context 传递信息：

Reviewer 3: ```suggestion
现在，你可以同时通过 props 和 context 传递信息：
```

</TasksContext.Provider>
```

相反，任何需要 tasks 的组件都可以从 `TaskContext` 中读取它：

```js {2}
export default function TaskList() {
  const tasks = useContext(TasksContext);
  // ...
```

> 要更新任务列表，任何组件都可以从 context 中读取 `dispatch` 函数并调用它：

Reviewer: ```suggestion
任何组件都可以从 context 中读取 `dispatch` 函数并调用它，从而更新任务列表：
```

    // ...
```

> **`TaskApp` 组件不会向下传递任何事件处理程序， `TaskList` 也不会。** 每个组件都会读取它需要的 context：

Reviewer 3: ```suggestion
**`TaskApp` 组件不会向下传递任何事件处理程序， `TaskList` 也不会**。每个组件都会读取它需要的 context：
```

</Sandpack>

> **state 仍然"存在于"顶层 `TaskBoard` 组件中，由 `useReducer` 进行管理。** 不过，组件树里的组件只要导入这些 context 之后就可以获取 `tasks` 和 `dispatch`。

Reviewer 3: ```suggestion
**state 仍然 “存在于” 顶层 `TaskBoard` 组件中，由 `useReducer` 进行管理**。不过，组件树里的组件只要导入这些 context 之后就可以获取 `tasks` 和 `dispatch`。
```

> ## 将所有的连接部分放入一个文件 {/*moving-all-wiring-into-a-single-file*/}

Reviewer: ```suggestion
## 将相关逻辑迁移到一个文件当中 {/*moving-all-wiring-into-a-single-file*/}
```

> 虽然不是必要的，但你可以通过将 reducer 和 context 移动到单个文件中来进一步整理组件。目前，"TasksContext.js"仅包含两个 context 声明：

Reviewer: ```suggestion
这不是必须的，但你可以通过将 reducer 和 context 移动到单个文件中来进一步整理组件。目前，“TasksContext.js” 仅包含两个 context 声明：
```

export const TasksDispatchContext = createContext(null);
```

> 这个文件即将变得臃肿！你将会把 reducer 移动到同一文件中。然后，你将在这个文件中声明一个新的 `TasksProvider` 组件。此组件将所有部分连接在一起：

Reviewer: 「这个文件即将变得臃肿！」
感觉这里翻译成「臃肿」不太合适，本意应该是「原本文件的逻辑很简单，现在会添加更多的内容。」

Reviewer 3: hmm... 的确，这个比较难处理。臃肿会让它听起来不大对劲……

Author: 直接不翻译这句如何?
「你将会把 reducer 移动到此文件中，然后声明一个新的 `TasksProvider` 组件。此组件将所有部分连接在一起：」
或者
「你将给这个文件添加更多内容，将 reducer 移动到此文件中，然后声明一个新的 `TasksProvider` 组件。此组件将所有部分连接在一起：」

Reviewer 3: 直接不翻译尽量不要，毕竟翻译还是要忠于原文的。原文是想调个皮……要不我们就“马上你的文件就要挤满代码啦！”这样？

Reviewer: > 直接不翻译这句如何? 「你将会把 reducer 移动到此文件中，然后声明一个新的 `TasksProvider` 组件。此组件将所有部分连接在一起：」 或者 「你将给这个文件添加更多内容，将 reducer 移动到此文件中，然后声明一个新的 `TasksProvider` 组件。此组件将所有部分连接在一起：」

我觉得后者 OK。

1. 它将管理 reducer 的状态。

> 2. 它将提供所有的 context 给组件树。

Reviewer: 这里「所有的 context」不是很明确，这里指的应该是上文的 “TasksContext” 和 “TasksDispatchContext”。另外 provide 直译为提供也有点奇怪，但我也没想到一个更合适的词...

Reviewer 3: 的确，both context 最好不要翻译成 `所有`。不过我觉得 `提供` 应该问题不大。

Author: "所有" 改为 "现有" 怎么样

Reviewer 3: @Ikaite 好主意！

> 3. 它将 [把子元素作为 prop](/learn/passing-props-to-a-component#passing-jsx-as-children)，所以你可以传递 JSX。

Reviewer: ```suggestion
3. 它将 [把 `children` 作为 prop](/learn/passing-props-to-a-component#passing-jsx-as-children)，所以你可以传递 JSX。
```

}
```

> **这将从你的 `TaskBoard` 组件中移除所有错综复杂的东西：**

Reviewer: ```suggestion
**这将会消除你的 `TaskBoard` 组件的复杂度：**
```

Author: "消除复杂度"感觉有点别扭,下面两种怎么样
1. **这将降低 `TaskBoard` 组件复杂度：**
2. **这将使 `TaskBoard` 组件更加直观：**

Reviewer 3: 主要它原文 all 这个词比较讨厌……我认为的确 @Ikaite 的听起来更好……

}
```

> 当一个组件需要读取 context，它可以通过以下函数实现：

Reviewer: ```suggestion
组件可以通过以下函数读取 context：
```

```js {5-7}
const tasks = useTasks();
const dispatch = useTasksDispatch();
```

> 这不会以任何方式改变行为，但它允许您稍后进一步分割这些 context 或向这些函数添加一些逻辑。**现在所有的 context 和 reducer 连接部分都在 `TasksContext.js`中。这保持了组件的干净和整洁，让我们专注于它们显示的内容，而不是它们从哪里获得数据：**

Reviewer 3: ```suggestion
这不会改变任何行为，但它会允许你之后进一步分割这些 context 或向这些函数添加一些逻辑。**现在所有的 context 和 reducer 连接部分都在 `TasksContext.js` 中。这保持了组件的干净和整洁，让我们专注于它们显示的内容，而不是它们从哪里获得数据：**
```

</Sandpack>

> 你可能会想 `TasksProvider` 作为页面的一部分是如何处理 tasks 的，`useTasks` 用来读取它们，`useTasksDispatch` 用来从组件树下的任何组件更新它们。

Reviewer: ```suggestion
你可以将 `TasksProvider` 视为页面的一部分，它知道如何处理 tasks。`useTasks` 用来读取它们，`useTasksDispatch` 用来从组件树下的任何组件更新它们。
```

> > 像 `useTasks` 和 `useTasksDispatch` 被称为 **[自定义钩子](/learn/reusing-logic-with-custom-hooks).** 如果你的函数名以 `use` 开头，它就被认为是一个自定义钩子。这让你可以使用其他 Hook，比如 `useContext`。

Reviewer: ```suggestion
> 像 `useTasks` 和 `useTasksDispatch` 这样的函数被称为 **[自定义 Hook](/learn/reusing-logic-with-custom-hooks)。** 如果你的函数名以 `use` 开头，它就被认为是一个自定义 Hook。这让你可以使用其他 Hook，比如 `useContext`。
```

> 随着应用的增长，你可能会有许多这样的 context 和 reducer 的组合。这是一种强大的方法，可以拓展你的应用并 [提升状态](/learn/sharing-state-between-components)，而无需在你想要访问组件树深处的数据时进行太多工作。

Reviewer: ```suggestion
随着应用的增长，你可能会有许多这样的 context 和 reducer 的组合。这是一种强大的拓展应用并 [提升状态](/learn/sharing-state-between-components) 的方式，让你在组件树深处访问数据时无需进行太多工作。
```

<Recap>

- 你可以将 reducer 与 context 相结合，让任何组件读取和更新它的状态。
- 为子组件提供 state 和 dispatch 函数：
  1. 创建两个 context (一个用于 state,一个用于 dispatch 函数)。
  2. 让组件的 context 使用 reducer。
  3. 使用组件中需要读取的 context。
- 你可以通过将所有传递信息的代码移动到单个文件中来进一步整理组件。
  - 你可以导出一个像 `TasksProvider` 可以提供 context 的组件。

>   - 你也可以导出自定义钩子像 `useTasks` 和 `useTasksDispatch`。

Reviewer: ```suggestion
  - 你也可以导出像 `useTasks` 和 `useTasksDispatch` 这样的自定义 Hook。
```

>   - 你也可以导出自定义钩子像 `useTasks` 和 `useTasksDispatch` 。

Reviewer 3: ```suggestion
  - 你也可以导出自定义钩子像 `useTasks` 和 `useTasksDispatch`。
```

全角字符与其他字符间不要空格。

> - 你可以在你的应用程序中大量配合使用 context 和 reducer。

Reviewer: ```suggestion
- 你可以在你的应用程序中大量使用 context 和 reducer 的组合。
```
